examples/arkode/C_serial/ark_analytic_partitioned.c

```diff
@@ -0,0 +1,263 @@
+static int f_linear(const sunrealtype t, const N_Vector y, N_Vector ydot,
+                    void* const user_data)
+{
+  sunrealtype lambda = ((UserData*)user_data)->lambda;
+  N_VScale(-lambda, y, ydot);
+  return 0;
+}
+
+/* RHS for f^2(t, y) = y^2 */
+static int f_nonlinear(const sunrealtype t, const N_Vector y, N_Vector ydot,
+                       void* const user_data)
+{
+  N_VProd(y, y, ydot);
+  return 0;
+}
+
+/* Compute the exact analytic solution */
+static N_Vector exact_sol(const N_Vector y0, const sunrealtype tf,
+                          const UserData* const user_data)
+{
+  N_Vector sol             = N_VClone(y0);
+  const sunrealtype y0_val = N_VGetArrayPointer(y0)[0];
+  const sunrealtype lambda = user_data->lambda;
+  N_VGetArrayPointer(sol)[0] =
+    lambda * y0_val / (y0_val - (y0_val - lambda) * SUNRexp(lambda * tf));
+  return sol;
+}
+
+/* Check function return value...
+    opt == 0 means SUNDIALS function allocates memory so check if
+             returned NULL pointer
+    opt == 1 means SUNDIALS function returns a flag so check if
+             flag >= 0
+    opt == 2 means function allocates memory so check if returned
+             NULL pointer
+*/
+static int check_flag(void* flagvalue, const char* funcname, int opt)
+{
+  int* errflag;
+
+  /* Check if SUNDIALS function returned NULL pointer - no memory allocated */
+  if (opt == 0 && flagvalue == NULL)
+  {
+    fprintf(stderr, "\nSUNDIALS_ERROR: %s() failed - returned NULL pointer\n\n",
+            funcname);
+    return 1;
+  }
+
+  /* Check if flag < 0 */
+  else if (opt == 1)
+  {
+    errflag = (int*)flagvalue;
+    if (*errflag < 0)
+    {
+      fprintf(stderr, "\nSUNDIALS_ERROR: %s() failed with flag = %d\n\n",
+              funcname, *errflag);
+      return 1;
+    }
+  }
+
+  /* Check if function returned NULL pointer - no memory allocated */
+  else if (opt == 2 && flagvalue == NULL)
+  {
+    fprintf(stderr, "\nMEMORY_ERROR: %s() failed - returned NULL pointer\n\n",
+            funcname);
+    return 1;
+  }
+
+  return 0;
+}
+
+int main(const int argc, char* const argv[])
+{
+  /* Parse arguments */
+  const char* const integrator_name = (argc > 1) ? argv[1] : "splitting";
+  if (strcmp(integrator_name, "splitting") != 0 &&
+      strcmp(integrator_name, "forcing") != 0)
+  {
+    fprintf(stderr, "Invalid integrator: %s\nMust be 'splitting' or 'forcing'\n",
+            integrator_name);
+    return 1;
+  }
+  const char* const coefficients_name = (argc > 2) ? argv[2] : NULL;
+
+  /* Problem parameters */
+  const sunrealtype t0           = SUN_RCONST(0.0);  /* initial time */
+  const sunrealtype tf           = SUN_RCONST(1.0);  /* final time */
+  const sunrealtype dt           = SUN_RCONST(0.01); /* outer time step */
+  const sunrealtype dt_linear    = dt / 5;  /* linear integrator time step */
+  const sunrealtype dt_nonlinear = dt / 10; /* nonlinear integrator time step */
+
+  UserData user_data = {.lambda = SUN_RCONST(2.0)};
```

Comment:
Since its just a C example, its fine, but beware that designated initializers like this one won't work with a C++ compiler unless using C++20 or later.